### ISSUE TITLE
S2-02 [Backend] XGBoost risk scorer — 47 features to 5 risk levels with confidence

### ISSUE THREAD

**Issue by NathanLuckock** [Nathan7108/HackathonPractice#16]
## Problem
This is the core prediction model. Without it, Sentinel has no risk scores. The XGBoost classifier takes the 47-feature vector from the pipeline and outputs a risk level (LOW/MODERATE/ELEVATED/HIGH/CRITICAL) with confidence probability. It's the number judges see on the map and the number GPT-4o explains.

## Goal
Build training dataset from ACLED fatalities + UCDP historical labels. Train XGBoost classifier on 47 features. Save model. Expose predict_risk(features) that returns risk level, score, confidence, and top drivers.

### In scope
- FEATURE_COLUMNS and RISK_LABELS constants (import from pipeline.py)
- build_training_dataset() — labels country-months from ACLED fatalities + UCDP intensity
- train_risk_scorer(training_df) — XGBoost with stratified split, classification report, feature importances
- predict_risk(features) -> {risk_level, risk_score, confidence, probabilities, top_drivers}
- Save models/risk_scorer.pkl and models/risk_label_encoder.pkl
- Print classification_report and top 10 features (MEMORIZE FOR JUDGES)

### Out of scope
- Automatic retraining (Sprint 3 tracker)
- FastAPI endpoint (Sprint 3)

## Acceptance Criteria
- [ ] build_training_dataset() produces labeled DataFrame with risk_label column
- [ ] Training labels include: LOW, MODERATE, ELEVATED, HIGH, CRITICAL
- [ ] train_risk_scorer() prints sklearn classification_report on test set
- [ ] train_risk_scorer() prints top 10 feature importances
- [ ] Models saved: models/risk_scorer.pkl, models/risk_label_encoder.pkl
- [ ] predict_risk(features_dict) returns dict with risk_level, risk_score (0-100), confidence (0-1), probabilities (dict), top_drivers (list of 5 feature names)
- [ ] predict_risk handles 47-feature dict from SentinelFeaturePipeline.compute()
- [ ] Running python -m backend.ml.risk_scorer builds dataset, trains, evaluates, tests prediction

### RISK_LABELS:
```python
RISK_LABELS = ['LOW', 'MODERATE', 'ELEVATED', 'HIGH', 'CRITICAL']
```

### FEATURE_COLUMNS: Import from pipeline
```python
from backend.ml.pipeline import FEATURE_COLUMNS
# 47 features — same list the pipeline outputs
```

### build_training_dataset():
The challenge is we don't have 47 features for every historical country-month. Strategy:

**Option A (Recommended for hackathon):** Build a synthetic/semi-real training set:
1. Load ACLED CSVs from data/acled/ for all 8 countries
2. Group by country + month
3. Compute as many features as possible per country-month (ACLED features are straightforward; GDELT features can be approximated; World Bank/sentiment filled with country-level averages or zeros)
4. Label using ACLED fatality thresholds:
   - fatalities >= 500 -> CRITICAL
   - fatalities >= 200 -> HIGH
   - fatalities >= 50 -> ELEVATED
   - fatalities >= 10 -> MODERATE
   - fatalities < 10 -> LOW
5. Also incorporate UCDP labels via build_ucdp_training_labels() from fetch_ucdp.py for historical extension

**The key insight:** Even with imperfect features for historical data, the model learns the relationship between conflict intensity and risk labels. At inference time, it gets all 47 real features.

```python
def build_training_dataset() -> pd.DataFrame:
    """
    Build labeled training dataset from ACLED + UCDP.
    Returns DataFrame with FEATURE_COLUMNS + 'risk_label'.
    """
    root = Path(__file__).resolve().parents[2]
    frames = []
    
    for code, info in MONITORED_COUNTRIES.items():
        acled_path = root / "data" / "acled" / f"{info['acled_name'].lower()}.csv"
        if not acled_path.exists():
            continue
        acled_df = pd.read_csv(acled_path)
        acled_df['event_date'] = pd.to_datetime(acled_df['event_date'], errors='coerce')
        acled_df['year_month'] = acled_df['event_date'].dt.to_period('M')
        
        for period, group in acled_df.groupby('year_month'):
            features = {}
            # Compute ACLED features for this month
            acled_features = compute_acled_features(group, window_days=9999)  # use all rows in group
            features.update(acled_features)
            
            # Fill remaining features with zeros (imperfect but functional)
            for col in FEATURE_COLUMNS:
                if col not in features:
                    features[col] = 0.0
            
            # Label from fatalities
            fatalities = float(group['fatalities'].astype(float).sum())
            if fatalities >= 500:
                features['risk_label'] = 'CRITICAL'
            elif fatalities >= 200:
                features['risk_label'] = 'HIGH'
            elif fatalities >= 50:
                features['risk_label'] = 'ELEVATED'
            elif fatalities >= 10:
                features['risk_label'] = 'MODERATE'
            else:
                features['risk_label'] = 'LOW'
            
            features['country_code'] = code
            frames.append(features)
    
    return pd.DataFrame(frames)
```

### train_risk_scorer():
```python
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.preprocessing import LabelEncoder

def train_risk_scorer(training_df: pd.DataFrame):
    le = LabelEncoder()
    le.fit(RISK_LABELS)
    
    X = training_df[FEATURE_COLUMNS].fillna(0)
    y = le.transform(training_df['risk_label'])
    
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42, stratify=y
    )
    
    model = xgb.XGBClassifier(
        n_estimators=500,
        max_depth=6,
        learning_rate=0.05,
        subsample=0.8,
        colsample_bytree=0.8,
        eval_metric='mlogloss',
        random_state=42,
        early_stopping_rounds=20,
    )
    model.fit(X_train, y_train, eval_set=[(X_test, y_test)], verbose=50)
    
    y_pred = model.predict(X_test)
    print(classification_report(y_test, y_pred, target_names=RISK_LABELS))
    
    importance = pd.Series(model.feature_importances_, index=FEATURE_COLUMNS)
    print("\nTop 10 Feature Importances:")
    print(importance.nlargest(10))
    
    joblib.dump(model, "models/risk_scorer.pkl")
    joblib.dump(le, "models/risk_label_encoder.pkl")
    return model
```

### predict_risk():
```python
def predict_risk(features: dict) -> dict:
    model = joblib.load("models/risk_scorer.pkl")
    le = joblib.load("models/risk_label_encoder.pkl")
    
    X = pd.DataFrame([{col: features.get(col, 0) for col in FEATURE_COLUMNS}])
    probabilities = model.predict_proba(X)[0]
    predicted_class = model.predict(X)[0]
    risk_level = le.inverse_transform([predicted_class])[0]
    
    importance = model.feature_importances_
    top_features = sorted(zip(FEATURE_COLUMNS, importance), key=lambda x: -x[1])[:5]
    
    return {
        "risk_level": risk_level,
        "risk_score": int(probabilities[predicted_class] * 100),
        "confidence": round(float(probabilities[predicted_class]), 3),
        "probabilities": dict(zip(RISK_LABELS, [round(float(p), 3) for p in probabilities])),
        "top_drivers": [f for f, _ in top_features],
    }
```

### Edge cases:
- Not all 5 risk classes may appear in training data (e.g., CRITICAL may be rare). Use stratify=y in train_test_split. If a class has < 2 samples, merge CRITICAL into HIGH.
- Some countries may have very few months of data. Include all available.
- XGBoost early_stopping_rounds=20 needs eval_set. If eval set is too small, reduce or remove.
- UCDP training labels: merge with build_ucdp_training_labels() if armed conflict CSV exists. Map UCDP's LOW/ELEVATED/HIGH to our 5-class system.

### CRITICAL for judges:
After training, WRITE DOWN:
- Test accuracy %
- Top 3 features
- Classification report (precision/recall per class)
You WILL be asked these numbers. Know them cold.

## Manual Test Steps
1. python -m backend.ml.risk_scorer — builds dataset, trains, prints report
2. Verify classification_report prints with all risk levels
3. Verify top 10 features printed
4. Verify models/risk_scorer.pkl and models/risk_label_encoder.pkl saved
5. Verify predict_risk() works with sample features from pipeline

---

## Cursor Agent Instructions

Repo: Nathan7108/HackathonPractice
Working directory: C:\Users\natha\HackathonPractice
Branch: feature/s2-02-xgboost-risk-scorer

### STEP 0 — Read context
```
cat PROGRESS.md
cat backend/ml/pipeline.py
cat backend/ml/data/fetch_acled.py
cat backend/ml/data/fetch_ucdp.py
```
Understand: FEATURE_COLUMNS (47 keys), MONITORED_COUNTRIES, compute_acled_features(), build_ucdp_training_labels()

### STEP 1 — Create branch
```
git checkout main && git pull origin main
git checkout -b feature/s2-02-xgboost-risk-scorer
```

### STEP 2 — Implement backend/ml/risk_scorer.py
Replace placeholder with full implementation:
1. Import FEATURE_COLUMNS from pipeline.py
2. RISK_LABELS = ['LOW', 'MODERATE', 'ELEVATED', 'HIGH', 'CRITICAL']
3. build_training_dataset() — ACLED monthly aggregation + labeling
4. train_risk_scorer(training_df) — XGBoost train, classification_report, save
5. predict_risk(features) — load model, predict, return dict
6. __main__ block: build dataset, print shape and label distribution, train, test predict_risk with sample

### STEP 3 — Validate
```
python -m backend.ml.risk_scorer
```
Expected: dataset built, model trained, classification report printed, models saved.

### STEP 4 — Commit and push
```
git add backend/ml/risk_scorer.py
git commit -m "feat: S2-02 XGBoost risk scorer — 47 features to 5 risk levels"
git push -u origin feature/s2-02-xgboost-risk-scorer
```

### STEP 5 — Open PR
```
gh pr create --title "S2-02 [Backend] XGBoost risk scorer" --body "Closes #16" --base main
```

### STEP 6 — Update PROGRESS.md
Record: accuracy %, top 3 features, training set size.

### Definition of Done
- [ ] build_training_dataset() produces labeled DataFrame
- [ ] train_risk_scorer() prints classification_report
- [ ] Top 10 feature importances printed
- [ ] models/risk_scorer.pkl and models/risk_label_encoder.pkl saved
- [ ] predict_risk() returns valid dict with risk_level, risk_score, confidence
- [ ] Pushed to origin, PR created

### If You Get Stuck
- FEATURE_COLUMNS: from backend.ml.pipeline import FEATURE_COLUMNS
- ACLED CSVs: data/acled/{country}.csv (lowercase country name)
- UCDP labels: from backend.ml.data.fetch_ucdp import build_ucdp_training_labels
- If stratify fails (too few samples in a class): merge CRITICAL into HIGH, or use stratify=None
- If early_stopping fails: remove early_stopping_rounds parameter
